python/ray/dag/communication_node.py

```diff
@@ -1,26 +1,199 @@
+from ray.experimental.channel.torch_tensor_type import (
+    Communicator,
+    TorchTensorType,
+)
 from ray.experimental.util.types import (
-    _CollectiveOp,
     AllGatherOp,
     AllReduceOp,
     ReduceScatterOp,
+    _CollectiveOp,
+    _P2POp,
 )
 from ray.util.annotations import DeveloperAPI
 
+import ray
+
+if TYPE_CHECKING:
+    import torch
+
+
+class _NcclOperation(ABC):
+    """
+    Represent metadata for a NCCL operation.
+    """
+
+    @abstractmethod
+    def execute(self, *args, **kwargs) -> Any:
+        """
+        Execute the NCCL operation in `ExecutableTask`.
+        """
+        raise NotImplementedError
+
+
+class _P2POperation(_NcclOperation):
+    """
+    Represent an executable NCCL P2P operation.
+    """
+
+    def __init__(self):
+        self._nccl_ch: Optional[ChannelInterface] = None
+
+    @property
+    def nccl_ch(self) -> Optional[ChannelInterface]:
+        return self._nccl_ch
+
+    @nccl_ch.setter
+    def nccl_ch(self, nccl_ch: ChannelInterface) -> None:
+        self._nccl_ch = nccl_ch
+
+
+class _P2PSendOperation(_P2POperation):
+    """
+    Represent an executable NCCL P2P send operation.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def execute(self, data: Any) -> None:
+        """
+        Execute the NCCL P2P send operation. Write the data via the NCCL channel.
+
+        Args:
+            data: The data to send.
+        """
+        self.nccl_ch.write(data)
+
 
-class _CollectiveOperation:
+class _P2PRecvOperation(_P2POperation):
+    """
+    Represent an executable NCCL P2P recv operation.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def execute(self) -> Any:
+        """
+        Execute the NCCL P2P recv operation. Read the data from the NCCL channel.
+
+        Return:
+            Data read from the NCCL channel.
+        """
+        return self.nccl_ch.read()
+
+
+@DeveloperAPI
+class P2PSendNode(ClassMethodNode):
+    """Represents a NCCL P2P send operation in a Ray DAG."""
+
+    def __init__(
+        self,
+        method_args: Tuple[ClassMethodNode],
+        other_args_to_resolve: Dict[str, Any],
+    ):
+        super().__init__(
+            method_name="p2p_send",
+            method_args=method_args,
+            method_kwargs=dict(),
+            method_options=dict(),
+            other_args_to_resolve=other_args_to_resolve,
+        )
+        self._send_op = _P2PSendOperation()
+
+        # Parse the input node.
+        if not (
+            isinstance(method_args, tuple)
+            and len(method_args) == 1
+            and isinstance(method_args[0], ClassMethodNode)
+        ):
+            raise ValueError("Expected a single input node as ClassMethodNode")
+        if isinstance(method_args[0], P2PSendNode) or isinstance(
+            method_args[0], P2PRecvNode
+        ):
+            raise ValueError("P2P send node cannot bind to another P2P node")
```

Comment:
Is it possible for users to write a graph that does this? If so, can you write a unit test that checks that the right error is thrown?